File: src/Java/myservicebus/src/main/java/com/myservicebus/InMemoryScheduleMessageProvider.java
package com.myservicebus;

import com.myservicebus.tasks.CancellationToken;
import java.time.Instant;
import java.util.UUID;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.function.Supplier;

public final class InMemoryScheduleMessageProvider implements ScheduleMessageProvider {
    private final PublishEndpoint publishEndpoint;
    private final SendEndpointProvider sendEndpointProvider;
    private final LocalDelayScheduler delayScheduler;
    private final InMemoryScheduledWorkSource source;
    private final Set<ScheduledWorkObserver> observers;

    public InMemoryScheduleMessageProvider(
            PublishEndpoint publishEndpoint,
            SendEndpointProvider sendEndpointProvider,
            LocalDelayScheduler delayScheduler) {
        this(publishEndpoint, sendEndpointProvider, delayScheduler, new InMemoryScheduledWorkSource(), Set.of());
    }

    public InMemoryScheduleMessageProvider(
            PublishEndpoint publishEndpoint,
            SendEndpointProvider sendEndpointProvider,
            LocalDelayScheduler delayScheduler,
            Set<ScheduledWorkObserver> observers) {
        this(publishEndpoint, sendEndpointProvider, delayScheduler, new InMemoryScheduledWorkSource(), observers);
    }

    public InMemoryScheduleMessageProvider(
            PublishEndpoint publishEndpoint,
            SendEndpointProvider sendEndpointProvider,
            LocalDelayScheduler delayScheduler,
            InMemoryScheduledWorkSource source,
            Set<ScheduledWorkObserver> observers) {
        this.publishEndpoint = publishEndpoint;
        this.sendEndpointProvider = sendEndpointProvider;
        this.delayScheduler = delayScheduler;
        this.source = source;
        this.observers = Set.copyOf(observers);
    }

    @Override
    public SchedulingDurability getDurability() {
        return SchedulingDurability.VOLATILE;
    }

    @Override
    public boolean supportsCancellation() {
        return true;
    }

    @Override
    public <T> CompletionStage<ScheduledMessageHandle> schedulePublish(
            Instant scheduledTime,
            T message,
            CancellationToken cancellationToken) {
        CompletableFuture<UUID> tokenReady = new CompletableFuture<>();
        return delayScheduler.schedule(scheduledTime, token -> tokenReady.thenCompose(tokenId ->
                        execute(tokenId, token, () -> publishEndpoint.publish(message, token))), cancellationToken)
                .thenApply(tokenId -> {
                    trackPending(tokenId, scheduledTime, message.getClass(), "Publish", null);
                    tokenReady.complete(tokenId);
                    return new ScheduledMessageHandle(tokenId, scheduledTime);
                });
    }

    @Override
    public <T> CompletionStage<ScheduledMessageHandle> scheduleSend(
            String destinationAddress,
            Instant scheduledTime,
            T message,
            CancellationToken cancellationToken) {
        CompletableFuture<UUID> tokenReady = new CompletableFuture<>();
        return delayScheduler.schedule(scheduledTime, token -> tokenReady.thenCompose(tokenId ->
                        execute(tokenId, token, () -> {
                            SendEndpoint endpoint = sendEndpointProvider.getSendEndpoint(destinationAddress);
                            return endpoint.send(message, token);
                        })), cancellationToken)
                .thenApply(tokenId -> {
                    trackPending(tokenId, scheduledTime, message.getClass(), "Send", destinationAddress);
                    tokenReady.complete(tokenId);
                    return new ScheduledMessageHandle(tokenId, scheduledTime);
                });
    }

    @Override
    public CompletionStage<ScheduleCancellationResult> cancel(UUID tokenId, CancellationToken cancellationToken) {
        return delayScheduler.cancel(tokenId).thenApply(cancelled -> {
            if (!cancelled) {
                return ScheduleCancellationResult.NOT_FOUND;
            }
            ScheduledWorkState state = source.remove(tokenId);
            if (state != null) {
                publish(new ScheduledWorkState(
                        state.tokenId(), state.provider(), state.durability(), state.workKind(), state.messageType(),
                        state.intent(), state.destinationAddress(), state.dueAtUtc(), ScheduledWorkStatus.CANCELLED,
                        "Cancelled", state.attempt(), Instant.now(), null));
            }
            return ScheduleCancellationResult.CANCELLED;
        });
    }

    private void trackPending(UUID tokenId, Instant scheduledTime, Class<?> messageType, String intent,
            String destinationAddress) {
        ScheduledWorkState state = new ScheduledWorkState(
                tokenId, "InMemory", getDurability(), "Message", messageType.getName(), intent,
                destinationAddress, scheduledTime, ScheduledWorkStatus.PENDING, "Pending", 0, Instant.now(), null);
        source.upsert(state);
        publish(state);
    }

    private CompletionStage<Void> execute(UUID tokenId, CancellationToken token,
            Supplier<CompletionStage<Void>> callback) {
        ScheduledWorkState state = source.get(tokenId);
        if (state == null) {
            return CompletableFuture.completedFuture(null);
        }
        ScheduledWorkState running = new ScheduledWorkState(
                state.tokenId(), state.provider(), state.durability(), state.workKind(), state.messageType(),
                state.intent(), state.destinationAddress(), state.dueAtUtc(), ScheduledWorkStatus.RUNNING,
                "Running", state.attempt() + 1, Instant.now(), null);
        source.upsert(running);
        publish(running);
        try {
            return callback.get().whenComplete((ignored, failure) -> {
                ScheduledWorkStatus status = failure == null ? ScheduledWorkStatus.COMPLETED : ScheduledWorkStatus.FAILED;
                String failureCategory = failure == null ? null : failure.getClass().getSimpleName();
                publish(new ScheduledWorkState(
                        running.tokenId(), running.provider(), running.durability(), running.workKind(),
                        running.messageType(), running.intent(), running.destinationAddress(), running.dueAtUtc(),
                        status, failure == null ? "Completed" : "Failed", running.attempt(), Instant.now(),
                        failureCategory));
                source.remove(tokenId);
            });
        } catch (RuntimeException exception) {
            publish(new ScheduledWorkState(
                    running.tokenId(), running.provider(), running.durability(), running.workKind(),
                    running.messageType(), running.intent(), running.destinationAddress(), running.dueAtUtc(),
                    ScheduledWorkStatus.FAILED, "Failed", running.attempt(), Instant.now(),
                    exception.getClass().getSimpleName()));
            source.remove(tokenId);
            return CompletableFuture.failedFuture(exception);
        }
    }

    private void publish(ScheduledWorkState state) {
        for (ScheduledWorkObserver observer : observers) {
            try {
                observer.observe(state);
            } catch (RuntimeException ignored) {
                // Scheduling must not depend on an optional observer.
            }
        }
    }
}

File: src/MyServiceBus/InMemoryScheduleMessageProvider.cs
namespace MyServiceBus;

public sealed class InMemoryScheduleMessageProvider : IScheduleMessageProvider
{
    private readonly IPublishEndpoint publishEndpoint;
    private readonly ISendEndpointProvider sendEndpointProvider;
    private readonly ILocalDelayScheduler delayScheduler;
    private readonly InMemoryScheduledWorkSource source;
    private readonly IReadOnlyList<IScheduledWorkObserver> observers;

    public InMemoryScheduleMessageProvider(
        IPublishEndpoint publishEndpoint,
        ISendEndpointProvider sendEndpointProvider,
        ILocalDelayScheduler delayScheduler)
        : this(publishEndpoint, sendEndpointProvider, delayScheduler, new InMemoryScheduledWorkSource(), [])
    {
    }

    public InMemoryScheduleMessageProvider(
        IPublishEndpoint publishEndpoint,
        ISendEndpointProvider sendEndpointProvider,
        ILocalDelayScheduler delayScheduler,
        IEnumerable<IScheduledWorkObserver>? observers)
        : this(publishEndpoint, sendEndpointProvider, delayScheduler, new InMemoryScheduledWorkSource(), observers ?? [])
    {
    }

    public InMemoryScheduleMessageProvider(
        IPublishEndpoint publishEndpoint,
        ISendEndpointProvider sendEndpointProvider,
        ILocalDelayScheduler delayScheduler,
        InMemoryScheduledWorkSource source,
        IEnumerable<IScheduledWorkObserver> observers)
    {
        this.publishEndpoint = publishEndpoint;
        this.sendEndpointProvider = sendEndpointProvider;
        this.delayScheduler = delayScheduler;
        this.source = source;
        this.observers = observers.ToArray();
    }

    public SchedulingDurability Durability => SchedulingDurability.Volatile;

    public bool SupportsCancellation => true;

    public async Task<ScheduledMessageHandle> SchedulePublish<T>(
        DateTime scheduledTime,
        T message,
        CancellationToken cancellationToken = default)
        where T : class
    {
        var tokenReady = new TaskCompletionSource<Guid>(TaskCreationOptions.RunContinuationsAsynchronously);
        var tokenId = await delayScheduler.Schedule(
            scheduledTime,
            async ct =>
            {
                var id = await tokenReady.Task.ConfigureAwait(false);
                await ExecuteAsync(id, ct, () => publishEndpoint.Publish(message, cancellationToken: ct)).ConfigureAwait(false);
            },
            cancellationToken);
        TrackPending(tokenId, scheduledTime, typeof(T), "Publish", null);
        tokenReady.SetResult(tokenId);
        return new ScheduledMessageHandle(tokenId, scheduledTime);
    }

    public async Task<ScheduledMessageHandle> ScheduleSend<T>(
        Uri destinationAddress,
        DateTime scheduledTime,
        T message,
        CancellationToken cancellationToken = default)
        where T : class
    {
        var tokenReady = new TaskCompletionSource<Guid>(TaskCreationOptions.RunContinuationsAsynchronously);
        async Task Callback(CancellationToken callbackCancellationToken)
        {
            var id = await tokenReady.Task.ConfigureAwait(false);
            await ExecuteAsync(id, callbackCancellationToken, async () =>
            {
                var endpoint = await sendEndpointProvider.GetSendEndpoint(destinationAddress);
                await endpoint.Send(message, cancellationToken: callbackCancellationToken);
            }).ConfigureAwait(false);
        }

        var tokenId = await delayScheduler.Schedule(scheduledTime, Callback, cancellationToken);
        TrackPending(tokenId, scheduledTime, typeof(T), "Send", destinationAddress.ToString());
        tokenReady.SetResult(tokenId);
        return new ScheduledMessageHandle(tokenId, scheduledTime);
    }

    public async Task<ScheduleCancellationResult> Cancel(Guid tokenId, CancellationToken cancellationToken = default)
    {
        if (!await delayScheduler.Cancel(tokenId))
            return ScheduleCancellationResult.NotFound;

        if (source.TryRemove(tokenId, out var state))
            Publish(state with
            {
                Status = ScheduledWorkStatus.Cancelled,
                ProviderStatus = "Cancelled",
                UpdatedAtUtc = DateTimeOffset.UtcNow
            });
        return ScheduleCancellationResult.Cancelled;
    }

    private void TrackPending(Guid tokenId, DateTime scheduledTime, Type messageType, string intent, string? destinationAddress)
    {
        var state = new ScheduledWorkState(
            tokenId,
            "InMemory",
            Durability,
            "Message",
            messageType.FullName ?? messageType.Name,
            intent,
            destinationAddress,
            new DateTimeOffset(scheduledTime.ToUniversalTime()),
            ScheduledWorkStatus.Pending,
            "Pending",
            0,
            DateTimeOffset.UtcNow);
        source.Upsert(state);
        Publish(state);
    }

    private async Task ExecuteAsync(Guid tokenId, CancellationToken cancellationToken, Func<Task> callback)
    {
        if (!source.TryGet(tokenId, out var state))
            return;

        var running = state with
        {
            Status = ScheduledWorkStatus.Running,
            ProviderStatus = "Running",
            Attempt = state.Attempt + 1,
            UpdatedAtUtc = DateTimeOffset.UtcNow
        };
        source.Upsert(running);
        Publish(running);
        try
        {
            await callback().ConfigureAwait(false);
            Publish(running with
            {
                Status = ScheduledWorkStatus.Completed,
                ProviderStatus = "Completed",
                UpdatedAtUtc = DateTimeOffset.UtcNow
            });
        }
        catch (Exception exception) when (exception is not OperationCanceledException || !cancellationToken.IsCancellationRequested)
        {
            Publish(running with
            {
                Status = ScheduledWorkStatus.Failed,
                ProviderStatus = "Failed",
                UpdatedAtUtc = DateTimeOffset.UtcNow,
                FailureCategory = exception.GetType().Name
            });
            throw;
        }
        finally
        {
            source.TryRemove(tokenId, out _);
        }
    }

    private void Publish(ScheduledWorkState state)
    {
        foreach (var observer in observers)
        {
            try
            {
                observer.Observe(state);
            }
            catch
            {
                // Scheduling must not depend on an optional observer.
            }
        }
    }
}
